pkg/vm/engine/tae/containers/vector.go

```diff
@@ -123,304 +284,221 @@ func (vec *vector[T]) Equals(o Vector) bool {
 
-	return
+	newDownstream := NewShallowCopyMoVecFromBytes(vec.GetType(), bs)
+	newDownstream.Nsp = nulls
+
+	vec.releaseDownstream()
+	vec.downstreamVector = newDownstream
 }
 
-func (vec *vector[T]) ReadFrom(r io.Reader) (n int64, err error) {
-	vec.releaseRoStorage()
-	var tmpn int64
-	// 1. Vector type
-	typeBuf := make([]byte, types.TSize)
-	if _, err = r.Read(typeBuf); err != nil {
-		return
-	}
-	vec.typ = types.DecodeType(typeBuf)
-	n += int64(len(typeBuf))
+// When a new Append() is happening on a SharedMemory vector, we allocate the data[] from the mpool.
+func (vec *vector[T]) tryPromoting() {
 
-	// 2. Nullable
-	oneBuf := make([]byte, 1)
-	if _, err = r.Read(oneBuf); err != nil {
-		return
-	}
-	nullable := types.DecodeFixed[bool](oneBuf)
-	n += 1
+	if !vec.isOwner {
+		src := vec.Bytes()
 
-	if nullable {
-		vec.impl = newNullableVecImpl(vec)
-	} else {
-		vec.impl = newVecImpl(vec)
-	}
+		// deep copy
+		newDownstream, _ := NewDeepCopyMoVecFromBytes(vec.GetType(), src, vec.GetAllocator())
+		newDownstream.Nsp = vec.downstreamVector.Nsp.Clone()
 
-	// 3. Data
-	if tmpn, err = vec.stlvec.ReadFrom(r); err != nil {
-		return
+		vec.downstreamVector = newDownstream
+		vec.isOwner = true
 	}
-	n += tmpn
+}
 
-	// 4. Null
-	if !nullable {
-		return
-	}
-	fourBuf := make([]byte, int(unsafe.Sizeof(uint32(0))))
-	if _, err = r.Read(fourBuf); err != nil {
-		return
-	}
-	n += int64(len(fourBuf))
-	nullSize := types.DecodeFixed[uint32](fourBuf)
-	if nullSize == 0 {
-		return
-	}
-	vec.nulls = roaring64.New()
-	if tmpn, err = vec.nulls.ReadFrom(r); err != nil {
+func (vec *vector[T]) ExtendWithOffset(src Vector, srcOff, srcLen int) {
+
+	if srcLen <= 0 {
 		return
 	}
-	n += tmpn
-	return
-}
 
-func (vec *vector[T]) ReadFromFile(f common.IVFile, buffer *bytes.Buffer) (err error) {
-	vec.releaseRoStorage()
-	stat := f.Stat()
-	var n []byte
-	var buf []byte
-	var tmpNode []byte
-	if stat.CompressAlgo() != compress.None {
-		osize := int(stat.OriginSize())
-		size := stat.Size()
-		tmpNode, err = vec.GetAllocator().Alloc(int(size))
-		if err != nil {
-			return
-		}
-		defer vec.GetAllocator().Free(tmpNode)
-		srcBuf := tmpNode[:size]
-		if _, err = f.Read(srcBuf); err != nil {
-			return
-		}
-		if buffer == nil {
-			n, err = vec.GetAllocator().Alloc(osize)
-			if err != nil {
-				return
-			}
-			buf = n[:osize]
-		} else {
-			buffer.Reset()
-			if osize > buffer.Cap() {
-				buffer.Grow(osize)
-			}
-			buf = buffer.Bytes()[:osize]
-		}
-		if _, err = compress.Decompress(srcBuf, buf, compress.Lz4); err != nil {
-			if n != nil {
-				vec.GetAllocator().Free(n)
-			}
-			return
-		}
+	// The downstream vector, ie CN vector needs isNull as argument.
+	// So, we can't directly call cn_vector.Append() without parsing the data.
+	// Hence, we are using src.Get(i) to retrieve the Null value as such from the src, and inserting
+	// it into the current CnVectorAdapter via this function.
+	for i := srcOff; i < srcOff+srcLen; i++ {
```

Comment:
I have some concerns about performance here.
Could we add an `Extend` API for CN-Vector? And it can be used here